File: android/app/src/main/java/com/jaber/cyclealarm/CycleAlarmService.java
package com.jaber.cyclealarm;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.AlarmManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.media.ToneGenerator;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.PowerManager;
import android.os.VibrationEffect;
import android.os.Vibrator;

public class CycleAlarmService extends Service {
    public static final String ACTION_START = "com.jaber.cyclealarm.START";
    public static final String ACTION_UPDATE = "com.jaber.cyclealarm.UPDATE";
    public static final String ACTION_PAUSE_RESUME = "com.jaber.cyclealarm.PAUSE_RESUME";
    public static final String ACTION_STOP = "com.jaber.cyclealarm.STOP";
    public static final String ACTION_PHASE_EXPIRED = "com.jaber.cyclealarm.PHASE_EXPIRED";
    public static final String EXTRA_DURATION_SECONDS = "durationSeconds";
    public static final String EXTRA_RING_SECONDS = "ringSeconds";
    public static final String EXTRA_TARGET = "target";
    public static final String EXTRA_TONE = "tone";
    public static final String EXTRA_LANGUAGE = "language";

    private static final String CHANNEL_ID = "cycle_alarm_running_v2";
    private static final int NOTIFICATION_ID = 42;
    private static final int PHASE_ALARM_REQUEST_CODE = 77;
    private static final String PREFS = "cycle_alarm_service";

    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Runnable ticker = this::tick;

    private PowerManager.WakeLock wakeLock;
    private ToneGenerator toneGenerator;
    private long phaseEndAt;
    private long remainingMs;
    private int durationSeconds = 300;
    private int ringSeconds = 5;
    private int target = 10;
    private int completed = 0;
    private String tone = "classic";
    private String language = "ar";
    private Mode mode = Mode.IDLE;
    private boolean completeAfterRing = false;

    private enum Mode {
        IDLE,
        RUNNING,
        PAUSED,
        RINGING,
        COMPLETE
    }

    @Override
    public void onCreate() {
        super.onCreate();
        createChannel();
        toneGenerator = new ToneGenerator(AudioManager.STREAM_ALARM, 100);
        loadState();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String action = intent == null ? null : intent.getAction();

        if (action == null) {
            restoreRunningService();
        } else if (ACTION_START.equals(action)) {
            startFromIntent(intent);
        } else if (ACTION_UPDATE.equals(action)) {
            updateFromIntent(intent);
        } else if (ACTION_PAUSE_RESUME.equals(action)) {
            pauseOrResume();
        } else if (ACTION_STOP.equals(action)) {
            stopAlarm();
        } else if (ACTION_PHASE_EXPIRED.equals(action)) {
            handlePhaseExpired();
        }

        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        handler.removeCallbacks(ticker);
        cancelPhaseAlarm();
        releaseWakeLock();
        if (toneGenerator != null) {
            toneGenerator.release();
            toneGenerator = null;
        }
        super.onDestroy();
    }

    private void startFromIntent(Intent intent) {
        durationSeconds = Math.max(1, intent.getIntExtra(EXTRA_DURATION_SECONDS, 300));
        ringSeconds = Math.max(1, intent.getIntExtra(EXTRA_RING_SECONDS, 5));
        target = Math.max(1, intent.getIntExtra(EXTRA_TARGET, 10));
        tone = intent.getStringExtra(EXTRA_TONE) == null ? "classic" : intent.getStringExtra(EXTRA_TONE);
        language = intent.getStringExtra(EXTRA_LANGUAGE) == null ? "ar" : intent.getStringExtra(EXTRA_LANGUAGE);
        completed = 0;
        completeAfterRing = false;
        acquireWakeLock();
        beginCountdown();
    }

    private void updateFromIntent(Intent intent) {
        long now = System.currentTimeMillis();
        long previousDurationMs = Math.max(1000L, durationSeconds * 1000L);
        long currentRemainingMs = mode == Mode.RUNNING
                ? Math.max(0L, phaseEndAt - now)
                : remainingMs;
        double remainingRatio = Math.max(0.0, Math.min(1.0, currentRemainingMs / (double) previousDurationMs));

        durationSeconds = Math.max(1, intent.getIntExtra(EXTRA_DURATION_SECONDS, durationSeconds));
        ringSeconds = Math.max(1, intent.getIntExtra(EXTRA_RING_SECONDS, ringSeconds));
        target = Math.max(1, intent.getIntExtra(EXTRA_TARGET, target));
        tone = intent.getStringExtra(EXTRA_TONE) == null ? tone : intent.getStringExtra(EXTRA_TONE);
        language = intent.getStringExtra(EXTRA_LANGUAGE) == null ? language : intent.getStringExtra(EXTRA_LANGUAGE);
        completed = Math.min(completed, target);
        completeAfterRing = completed >= target;

        if (mode == Mode.RUNNING) {
            remainingMs = Math.max(1000L, Math.round(durationSeconds * 1000L * remainingRatio));
            phaseEndAt = now + remainingMs;
            schedulePhaseAlarm(phaseEndAt);
            scheduleTick();
        } else if (mode == Mode.PAUSED) {
            remainingMs = Math.max(1000L, Math.round(durationSeconds * 1000L * remainingRatio));
        }

        saveState();

        if (mode != Mode.IDLE) {
            startForeground(NOTIFICATION_ID, buildNotification());
        }
    }

    private void beginCountdown() {
        mode = Mode.RUNNING;
        remainingMs = durationSeconds * 1000L;
        phaseEndAt = System.currentTimeMillis() + remainingMs;
        saveState();
        startForeground(NOTIFICATION_ID, buildNotification());
        schedulePhaseAlarm(phaseEndAt);
        scheduleTick();
    }

    private void finishCycle() {
        completed += 1;
        completeAfterRing = completed >= target;
        mode = Mode.RINGING;
        phaseEndAt = System.currentTimeMillis() + ringSeconds * 1000L;
        saveState();
        playRingTone();
        vibrate();
        startForeground(NOTIFICATION_ID, buildNotification());
        schedulePhaseAlarm(phaseEndAt);
        scheduleTick();
    }

    private void completeTarget() {
        mode = Mode.COMPLETE;
        handler.removeCallbacks(ticker);
        cancelPhaseAlarm();
        saveState();
        playCompleteTone();
        vibrate();
        releaseWakeLock();
        startForeground(NOTIFICATION_ID, buildNotification());
    }

    private void tick() {
        long now = System.currentTimeMillis();

        if (mode == Mode.RUNNING) {
            remainingMs = Math.max(0, phaseEndAt - now);
            if (remainingMs <= 0) {
                finishCycle();
                return;
            }
        }

        if (mode == Mode.RINGING) {
            if (now >= phaseEndAt) {
                if (completeAfterRing) {
                    completeTarget();
                } else {
                    beginCountdown();
                }
                return;
            }
            playRingTone();
            vibrate();
        }

        startForeground(NOTIFICATION_ID, buildNotification());
        scheduleTick();
    }

    private void scheduleTick() {
        handler.removeCallbacks(ticker);
        handler.postDelayed(ticker, mode == Mode.RINGING ? 900L : 1000L);
    }

    private void handlePhaseExpired() {
        acquireWakeLock();

        if (mode == Mode.RUNNING) {
            finishCycle();
            return;
        }

        if (mode == Mode.RINGING) {
            if (completeAfterRing) {
                completeTarget();
            } else {
                beginCountdown();
            }
        }
    }

    private void pauseOrResume() {
        if (mode == Mode.RUNNING || mode == Mode.RINGING) {
            remainingMs = Math.max(0, phaseEndAt - System.currentTimeMillis());
            mode = Mode.PAUSED;
            saveState();
            startForeground(NOTIFICATION_ID, buildNotification());
            handler.removeCallbacks(ticker);
            cancelPhaseAlarm();
            return;
        }

        if (mode == Mode.PAUSED) {
            mode = Mode.RUNNING;
            phaseEndAt = System.currentTimeMillis() + Math.max(1000L, remainingMs);
            saveState();
            startForeground(NOTIFICATION_ID, buildNotification());
            schedulePhaseAlarm(phaseEndAt);
            scheduleTick();
        }
    }

    private void stopAlarm() {
        handler.removeCallbacks(ticker);
        cancelPhaseAlarm();
        clearState();
        releaseWakeLock();
        stopForeground(true);
        stopSelf();
    }

    private Notification buildNotification() {
        Intent openIntent = new Intent(this, MainActivity.class);
        PendingIntent openPendingIntent = PendingIntent.getActivity(
                this,
                1,
                openIntent,
                pendingIntentFlags()
        );

        PendingIntent pauseIntent = serviceAction(2, ACTION_PAUSE_RESUME);
        PendingIntent stopIntent = serviceAction(3, ACTION_STOP);

        String title = text("المنبه يعمل", "Alarm çalışıyor", "Alarm running");
        String status;
        if (mode == Mode.RUNNING) {
            status = text("الدورة ", "Tur ", "Cycle ") + (completed + 1) + " / " + target + " - " + formatTime(remainingMs);
        } else if (mode == Mode.RINGING) {
            status = text("تنبيه الدورة ", "Tur alarmı ", "Cycle alarm ") + completed + " / " + target;
        } else if (mode == Mode.PAUSED) {
            status = text("متوقف مؤقتًا - المكتمل ", "Duraklatıldı - tamamlanan ", "Paused - completed ") + completed + " / " + target;
        } else if (mode == Mode.COMPLETE) {
            title = text("اكتمل العدد", "Sayı tamamlandı", "Count complete");
            status = completed + " / " + target;
        } else {
            status = "";
        }

        Notification.Builder builder = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                ? new Notification.Builder(this, CHANNEL_ID)
                : new Notification.Builder(this);

        builder
                .setSmallIcon(R.drawable.ic_launcher)
                .setContentTitle(title)
                .setContentText(status)
                .setContentIntent(openPendingIntent)
                .setOngoing(mode != Mode.COMPLETE)
                .setOnlyAlertOnce(mode != Mode.RINGING && mode != Mode.COMPLETE)
                .setCategory(Notification.CATEGORY_ALARM)
                .setPriority(Notification.PRIORITY_HIGH)
                .setVisibility(Notification.VISIBILITY_PUBLIC)
                .setShowWhen(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && mode == Mode.RUNNING) {
            builder.setWhen(phaseEndAt);
            builder.setUsesChronometer(true);
            builder.setChronometerCountDown(true);
        }

        if (mode != Mode.COMPLETE) {
            builder.addAction(R.drawable.ic_launcher, mode == Mode.PAUSED ? text("متابعة", "Devam", "Resume") : text("إيقاف مؤقت", "Duraklat", "Pause"), pauseIntent);
            builder.addAction(R.drawable.ic_launcher, text("إيقاف", "Durdur", "Stop"), stopIntent);
        } else {
            builder.addAction(R.drawable.ic_launcher, text("إيقاف", "Durdur", "Stop"), stopIntent);
        }

        return builder.build();
    }

    private PendingIntent serviceAction(int requestCode, String action) {
        Intent intent = new Intent(this, CycleAlarmService.class);
        intent.setAction(action);
        return PendingIntent.getService(this, requestCode, intent, pendingIntentFlags());
    }

    private int pendingIntentFlags() {
        int flags = PendingIntent.FLAG_UPDATE_CURRENT;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            flags |= PendingIntent.FLAG_IMMUTABLE;
        }
        return flags;
    }

    private void createChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            return;
        }

        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "Cycle Alarm",
                NotificationManager.IMPORTANCE_HIGH
        );
        channel.setDescription("Persistent alarm timer and lock screen controls");
        channel.setLockscreenVisibility(Notification.VISIBILITY_PUBLIC);
        channel.enableVibration(true);
        NotificationManager manager = getSystemService(NotificationManager.class);
        manager.createNotificationChannel(channel);
    }

    private void schedulePhaseAlarm(long triggerAtMillis) {
        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        PendingIntent pendingIntent = phaseAlarmIntent();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && !alarmManager.canScheduleExactAlarms()) {
            alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
        } else {
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
        }
    }

    private void restoreRunningService() {
        if (mode == Mode.IDLE || mode == Mode.COMPLETE) {
            return;
        }

        acquireWakeLock();
        startForeground(NOTIFICATION_ID, buildNotification());

        if (mode == Mode.RUNNING || mode == Mode.RINGING) {
            schedulePhaseAlarm(phaseEndAt);
            scheduleTick();
        }
    }

    private void saveState() {
        getSharedPreferences(PREFS, MODE_PRIVATE)
                .edit()
                .putString("mode", mode.name())
                .putLong("phaseEndAt", phaseEndAt)
                .putLong("remainingMs", remainingMs)
                .putInt("durationSeconds", durationSeconds)
                .putInt("ringSeconds", ringSeconds)
                .putInt("target", target)
                .putInt("completed", completed)
                .putString("tone", tone)
                .putString("language", language)
                .putBoolean("completeAfterRing", completeAfterRing)
                .apply();
    }

    private void loadState() {
        SharedPreferences prefs = getSharedPreferences(PREFS, MODE_PRIVATE);
        String savedMode = prefs.getString("mode", Mode.IDLE.name());

        try {
            mode = Mode.valueOf(savedMode);
        } catch (IllegalArgumentException ignored) {
            mode = Mode.IDLE;
        }

        phaseEndAt = prefs.getLong("phaseEndAt", 0L);
        remainingMs = prefs.getLong("remainingMs", durationSeconds * 1000L);
        durationSeconds = prefs.getInt("durationSeconds", durationSeconds);
        ringSeconds = prefs.getInt("ringSeconds", ringSeconds);
        target = prefs.getInt("target", target);
        completed = prefs.getInt("completed", completed);
        tone = prefs.getString("tone", tone);
        language = prefs.getString("language", language);
        completeAfterRing = prefs.getBoolean("completeAfterRing", false);
    }

    private void clearState() {
        getSharedPreferences(PREFS, MODE_PRIVATE).edit().clear().apply();
        mode = Mode.IDLE;
    }

    private void cancelPhaseAlarm() {
        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        alarmManager.cancel(phaseAlarmIntent());
    }

    private PendingIntent phaseAlarmIntent() {
        Intent intent = new Intent(this, CycleAlarmService.class);
        intent.setAction(ACTION_PHASE_EXPIRED);
        return PendingIntent.getService(this, PHASE_ALARM_REQUEST_CODE, intent, pendingIntentFlags());
    }

    private void acquireWakeLock() {
        if (wakeLock != null && wakeLock.isHeld()) {
            return;
        }

        PowerManager powerManager = (PowerManager) getSystemService(Context.POWER_SERVICE);
        wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "CycleAlarm:Timer");
        wakeLock.acquire();
    }

    private void releaseWakeLock() {
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
        }
        wakeLock = null;
    }

    private void playRingTone() {
        if (toneGenerator == null) {
            return;
        }

        if ("soft".equals(tone)) {
            toneGenerator.startTone(ToneGenerator.TONE_PROP_BEEP, 350);
        } else if ("urgent".equals(tone)) {
            toneGenerator.startTone(ToneGenerator.TONE_CDMA_ALERT_CALL_GUARD, 550);
        } else {
            toneGenerator.startTone(ToneGenerator.TONE_CDMA_ALERT_NETWORK_LITE, 450);
        }
    }

    private void playCompleteTone() {
        if (toneGenerator != null) {
            toneGenerator.startTone(ToneGenerator.TONE_CDMA_ALERT_AUTOREDIAL_LITE, 1200);
        }
    }

    private void vibrate() {
        Vibrator vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
        if (vibrator == null || !vibrator.hasVibrator()) {
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(450, VibrationEffect.DEFAULT_AMPLITUDE));
        } else {
            vibrator.vibrate(450);
        }
    }

    private String formatTime(long milliseconds) {
        long totalSeconds = Math.max(0, milliseconds / 1000L);
        long minutes = totalSeconds / 60L;
        long seconds = totalSeconds % 60L;
        return String.format("%02d:%02d", minutes, seconds);
    }

    private String text(String arabic, String turkish) {
        return "tr".equals(language) ? turkish : arabic;
    }

    private String text(String arabic, String turkish, String english) {
        if ("en".equals(language)) {
            return english;
        }
        return "tr".equals(language) ? turkish : arabic;
    }
}

File: android/app/src/main/java/com/kallaa/cyclealarm/CycleAlarmService.kt
// المسار المطلوب: android/app/src/main/java/com/kallaa/cyclealarm/CycleAlarmService.kt
package com.kallaa.cyclealarm

import android.app.AlarmManager
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.media.RingtoneManager
import android.os.Build
import android.os.IBinder
import android.os.VibrationEffect
import android.os.Vibrator
import androidx.core.app.NotificationCompat
import org.json.JSONObject

class CycleAlarmService : Service() {

    private var mediaPlayer: MediaPlayer? = null
    private var vibrator: Vibrator? = null

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        createNotificationChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val action = intent?.action ?: ACTION_KEEP_ALIVE
        when (action) {
            ACTION_START -> handleStart(intent)
            ACTION_PAUSE_RESUME -> handlePauseResume()
            ACTION_STOP -> handleStop()
            ACTION_UPDATE -> handleUpdate(intent)
            ACTION_PHASE_END -> handlePhaseEnd()
        }
        return START_STICKY
    }

    // ---------- إجراءات عامة ----------

    private fun handleStart(intent: Intent?) {
        val durationSeconds = intent?.getIntExtra(EXTRA_DURATION, 300) ?: 300
        val ringSeconds = intent?.getIntExtra(EXTRA_RING, 5) ?: 5
        val target = intent?.getIntExtra(EXTRA_TARGET, 10) ?: 10
        val tone = intent?.getStringExtra(EXTRA_TONE) ?: "classic"
        val language = intent?.getStringExtra(EXTRA_LANG) ?: "ar"

        stopRinging()
        cancelAlarm()

        val prefs = prefs()
        prefs.edit()
            .putString(KEY_MODE, "RUNNING")
            .putInt(KEY_DURATION, durationSeconds)
            .putInt(KEY_RING, ringSeconds)
            .putInt(KEY_TARGET, target)
            .putString(KEY_TONE, tone)
            .putString(KEY_LANG, language)
            .putInt(KEY_COMPLETED, 0)
            .putBoolean(KEY_COMPLETE_AFTER_RING, false)
            .putLong(KEY_PHASE_END, System.currentTimeMillis() + durationSeconds * 1000L)
            .putLong(KEY_REMAINING, 0)
            .apply()

        scheduleAlarm(prefs.getLong(KEY_PHASE_END, 0))
        startForeground(NOTIFICATION_ID, buildNotification(notifTextRunning()))
    }

    private fun handlePauseResume() {
        val prefs = prefs()
        val mode = prefs.getString(KEY_MODE, "IDLE")
        val now = System.currentTimeMillis()

        when (mode) {
            "RUNNING" -> {
                val remaining = (prefs.getLong(KEY_PHASE_END, now) - now).coerceAtLeast(0)
                cancelAlarm()
                prefs.edit()
                    .putString(KEY_MODE, "PAUSED")
                    .putLong(KEY_REMAINING, remaining)
                    .apply()
                updateNotification(notifTextPaused())
            }
            "RINGING" -> {
                val remaining = (prefs.getLong(KEY_PHASE_END, now) - now).coerceAtLeast(1000)
                cancelAlarm()
                stopRinging()
                prefs.edit()
                    .putString(KEY_MODE, "PAUSED")
                    .putLong(KEY_REMAINING, remaining)
                    .apply()
                updateNotification(notifTextPaused())
            }
            "PAUSED" -> {
                val remaining = prefs.getLong(KEY_REMAINING, 0)
                val wasRinging = remaining in 1..prefs.getInt(KEY_RING, 5) * 1000L
                val newMode = if (wasRinging) "RINGING" else "RUNNING"
                val phaseEnd = now + remaining.coerceAtLeast(1000)
                prefs.edit()
                    .putString(KEY_MODE, newMode)
                    .putLong(KEY_PHASE_END, phaseEnd)
                    .apply()
                scheduleAlarm(phaseEnd)
                if (newMode == "RINGING") startRinging()
                updateNotification(notifTextRunning())
            }
        }
    }

    private fun handleStop() {
        cancelAlarm()
        stopRinging()
        prefs().edit().putString(KEY_MODE, "IDLE").apply()
        stopForeground(true)
        stopSelf()
    }

    private fun handleUpdate(intent: Intent?) {
        val prefs = prefs()
        intent?.getIntExtra(EXTRA_DURATION, -1)?.takeIf { it > 0 }?.let { prefs.edit().putInt(KEY_DURATION, it).apply() }
        intent?.getIntExtra(EXTRA_RING, -1)?.takeIf { it > 0 }?.let { prefs.edit().putInt(KEY_RING, it).apply() }
        intent?.getIntExtra(EXTRA_TARGET, -1)?.takeIf { it > 0 }?.let { prefs.edit().putInt(KEY_TARGET, it).apply() }
        intent?.getStringExtra(EXTRA_TONE)?.let { prefs.edit().putString(KEY_TONE, it).apply() }
        intent?.getStringExtra(EXTRA_LANG)?.let { prefs.edit().putString(KEY_LANG, it).apply() }
    }

    private fun handlePhaseEnd() {
        val prefs = prefs()
        val mode = prefs.getString(KEY_MODE, "IDLE")
        val now = System.currentTimeMillis()

        if (mode == "RUNNING") {
            val completed = prefs.getInt(KEY_COMPLETED, 0) + 1
            val target = prefs.getInt(KEY_TARGET, 10)
            val ringSeconds = prefs.getInt(KEY_RING, 5)
            val completeAfterRing = completed >= target

            prefs.edit()
                .putInt(KEY_COMPLETED, completed)
                .putBoolean(KEY_COMPLETE_AFTER_RING, completeAfterRing)
                .putString(KEY_MODE, "RINGING")
                .putLong(KEY_PHASE_END, now + ringSeconds * 1000L)
                .apply()

            scheduleAlarm(now + ringSeconds * 1000L)
            startRinging()
            updateNotification(notifTextRinging())
        } else if (mode == "RINGING") {
            stopRinging()
            val completeAfterRing = prefs.getBoolean(KEY_COMPLETE_AFTER_RING, false)

            if (completeAfterRing) {
                prefs.edit().putString(KEY_MODE, "COMPLETE").apply()
                showCompleteNotification()
                stopForeground(true)
                stopSelf()
            } else {
                val durationSeconds = prefs.getInt(KEY_DURATION, 300)
                val phaseEnd = now + durationSeconds * 1000L
                prefs.edit()
                    .putString(KEY_MODE, "RUNNING")
                    .putLong(KEY_PHASE_END, phaseEnd)
                    .apply()
                scheduleAlarm(phaseEnd)
                updateNotification(notifTextRunning())
            }
        }
    }

    // ---------- منبه النظام (رنة الهاتف الحقيقية) ----------

    private fun startRinging() {
        stopRinging()
        try {
            val uri = RingtoneManager.getActualDefaultRingtoneUri(this, RingtoneManager.TYPE_ALARM)
                ?: RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM)

            mediaPlayer = MediaPlayer().apply {
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ALARM)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                        .build()
                )
                setDataSource(this@CycleAlarmService, uri)
                isLooping = true
                prepare()
                start()
            }
        } catch (_: Exception) {
            // تجاهل: بعض الأجهزة قد تمنع تشغيل صوت المنبه في ظروف معينة
        }

        val pattern = longArrayOf(0, 500, 300, 500, 300)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator?.vibrate(VibrationEffect.createWaveform(pattern, 0))
        } else {
            @Suppress("DEPRECATION")
            vibrator?.vibrate(pattern, 0)
        }
    }

    private fun stopRinging() {
        try {
            mediaPlayer?.stop()
            mediaPlayer?.release()
        } catch (_: Exception) { /* توقف بالفعل */ }
        mediaPlayer = null
        vibrator?.cancel()
    }

    // ---------- AlarmManager ----------

    private fun scheduleAlarm(triggerAtMillis: Long) {
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent = phaseEndPendingIntent()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent)
        } else {
            alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent)
        }
    }

    private fun cancelAlarm() {
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarmManager.cancel(phaseEndPendingIntent())
    }

    private fun phaseEndPendingIntent(): PendingIntent {
        val intent = Intent(this, AlarmReceiver::class.java).setAction(ACTION_PHASE_END)
        val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        else PendingIntent.FLAG_UPDATE_CURRENT
        return PendingIntent.getBroadcast(this, REQUEST_CODE, intent, flags)
    }

    // ---------- إشعارات ----------

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(CHANNEL_ID, "Cycle Timer", NotificationManager.IMPORTANCE_LOW)
            (getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager).createNotificationChannel(channel)
        }
    }

    private fun buildNotification(text: String): Notification {
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("Kallaa Cycle Timer")
            .setContentText(text)
            .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
            .setOngoing(true)
            .build()
    }

    private fun updateNotification(text: String) {
        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.notify(NOTIFICATION_ID, buildNotification(text))
    }

    private fun showCompleteNotification() {
        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("Kallaa Cycle Timer")
            .setContentText(notifTextComplete())
            .setSmallIcon(android.R.drawable.ic_lock_idle_alarm)
            .setAutoCancel(true)
            .build()
        manager.notify(NOTIFICATION_ID + 1, notification)
    }

    private fun notifTextRunning() = "العداد يعمل • ${prefs().getInt(KEY_COMPLETED, 0)}/${prefs().getInt(KEY_TARGET, 10)}"
    private fun notifTextPaused() = "متوقف مؤقتًا"
    private fun notifTextRinging() = "تنبيه — انتهت دورة"
    private fun notifTextComplete() = "اكتمل العدد المطلوب"

    private fun prefs(): SharedPreferences = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    companion object {
        const val PREFS_NAME = "cycle_alarm_state"
        const val CHANNEL_ID = "cycle_alarm_channel"
        const val NOTIFICATION_ID = 1001
        const val REQUEST_CODE = 2002

        const val ACTION_START = "com.kallaa.cyclealarm.START"
        const val ACTION_PAUSE_RESUME = "com.kallaa.cyclealarm.PAUSE_RESUME"
        const val ACTION_STOP = "com.kallaa.cyclealarm.STOP"
        const val ACTION_UPDATE = "com.kallaa.cyclealarm.UPDATE"
        const val ACTION_PHASE_END = "com.kallaa.cyclealarm.PHASE_END"
        const val ACTION_KEEP_ALIVE = "com.kallaa.cyclealarm.KEEP_ALIVE"

        const val EXTRA_DURATION = "durationSeconds"
        const val EXTRA_RING = "ringSeconds"
        const val EXTRA_TARGET = "target"
        const val EXTRA_TONE = "tone"
        const val EXTRA_LANG = "language"

        const val KEY_MODE = "mode"
        const val KEY_DURATION = "durationSeconds"
        const val KEY_RING = "ringSeconds"
        const val KEY_TARGET = "target"
        const val KEY_TONE = "tone"
        const val KEY_LANG = "language"
        const val KEY_COMPLETED = "completed"
        const val KEY_COMPLETE_AFTER_RING = "completeAfterRing"
        const val KEY_PHASE_END = "phaseEndAt"
        const val KEY_REMAINING = "remainingMs"

        /** يقرأ الحالة الحالية كنص JSON بنفس الصيغة التي يتوقعها app.js في applyAlarmSnapshot() */
        fun readStateJson(context: Context): String {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = JSONObject()
            json.put("mode", prefs.getString(KEY_MODE, "IDLE"))
            json.put("durationSeconds", prefs.getInt(KEY_DURATION, 300))
            json.put("ringSeconds", prefs.getInt(KEY_RING, 5))
            json.put("target", prefs.getInt(KEY_TARGET, 10))
            json.put("tone", prefs.getString(KEY_TONE, "classic"))
            json.put("language", prefs.getString(KEY_LANG, "ar"))
            json.put("completed", prefs.getInt(KEY_COMPLETED, 0))
            json.put("completeAfterRing", prefs.getBoolean(KEY_COMPLETE_AFTER_RING, false))
            json.put("phaseEndAt", prefs.getLong(KEY_PHASE_END, 0))
            json.put("remainingMs", prefs.getLong(KEY_REMAINING, 0))
            return json.toString()
        }
    }
}
